Copy channelhandle by value instead of pointers

ywwg opened the pull request:
Store channelhandles in messages by value instead of pointers. This fixes issues with pointers to deleted channelhandles being accessed, causing crashes, mostly with QT6.
    
Converting to values means we can no longer trivially initialize the union/struct by memsetting all of the subtypes, because two of the struts are no longer trivial.
    
Instead, we memset what we think is the largest datatype. SetEffectChainParameters includes a bool, an enum, and a double, so it's the largest one.

ywwg committed:
Store channelhandles in messages by value instead of pointers. This fixes issues with pointers to deleted channelhandles being accessed, causing crashes, mostly with QT6.

Converting to values means we can no longer trivially initialize the union/struct by memsetting all of the subtypes, because two of the struts are no longer trivial.

Instead, we memset what we think is the largest datatype. SetEffectChainParameters includes a bool, an enum, and a double, so it's the largest one.

ywwg commented:
> This failure was expected. There is a reason why pointers are used.

From your other post it sounds like you're unwilling to explain the reasons so I'm not sure what to do with this information

uklotzde commented on src/engine/effects/message.h:
```diff
@@ -97,10 +95,10 @@ struct EffectsRequest {
         struct {
             EffectStatesMapArray* pEffectStatesMapArray;
-            const ChannelHandle* pChannelHandle;
+            ChannelHandle channelHandle;
```
I was expecting that ChannelHandle with its custom constructor is not usable in this union. Removing the memset seem to work, though.

Be-ing replied:
That seems really brittle and easy to break when refactoring.

ywwg replied:
agree -- for now we think it's just better to state that we don't guarantee initialization and callers are responsible for it.  We can clean this up with the followup rewrite using variant or whatever

uklotzde commented on src/engine/effects/engineeffectchain.h:
```diff
@@ -64,9 +64,9 @@ class EngineEffectChain final : public EffectsRequestHandler {
     bool addEffect(EngineEffect* pEffect, int iIndex);
     bool removeEffect(EngineEffect* pEffect, int iIndex);
-    bool enableForInputChannel(const ChannelHandle* inputHandle,
+    bool enableForInputChannel(const ChannelHandle inputHandle,
```
Please remove the `const` qualifiers from all by-value arguments.

ywwg replied:
I believe as long as we call memset on the largest struct item, that will be sufficient for zeroing out the data

ywwg commented:
I'm going to squash this down since there are some bad intermediate commits

daschuer commented on src/engine/effects/message.h:
```diff
@@ -40,17 +40,9 @@ struct EffectsRequest {
-#undef CLEAR_STRUCT
+        // zero out the struct with the largest size to ensure the entire union memory is set to
+        // zero.
+        memset(&SetEffectChainParameters, 0, sizeof(SetEffectChainParameters));
```
Can we name the anonymous union here and then clear the memory by sizeof(union Payload)?
All other ideas I have in mind are breaking during refectoring. 

However, I actually think that a zeroed memory is in the same way dangerous as a ununutalized memory.
We must not use it in both cases. It looks like the rest of the control structures take already care of it. So if zeroing has an effect, it will only obfusticate an issue we have anyway. 

Thats why I would prefere to not initialize the memory. It makes more sense to asserrt, the validity of the control structures and introduce constructors, that take care of them in one call and cannot fail.

daschuer commented:
While this is already a good way forward which is already mergable, I have not yet understand how it avoids the bad_allic exception. Do we have an evidence that this fixes the issue?

ywwg commented:
> While this is already a good way forward which is already mergable, I have not yet understand how it avoids the bad_allic exception. Do we have an evidence that this fixes the issue?

Yes.  We traced the bad alloc to the following:

* at https://github.com/mixxxdj/mixxx/blob/main/src/effects/effectchain.cpp#L397 we assign the pointer in a message by taking the address of the channelhandle member.
* When that pointer is later accessed for use in resizing the channelhandle map, somewhere along the way the original thing being pointed to has been deleted / moved.  
* The memory is invalid, and we request a humongous iSize, causing a bad alloc

After this change, even running the UnEjectTest many times does not crash, whereas it crashed about 50% of the time before

ywwg commented:
I think you're right, zeroing out data won't help if we're not initializing correctly anyway so it doesn't really buy us anything.

ywwg committed:
EffectsRequest: Remove union initialization.

Add comment to make clear callers are responsible for initializing the values correctly.
Double-checked call points to make sure all creations were correct.
Removed unused struct parameter that I found during this process.

daschuer reviewed (approved):
Thank you. Merge?

ywwg commented:
I'd like to give @Be-ing a chance to chime in -- we plan to rewrite this subsystem so this is just a stop-gap fix

daschuer commented:
@ywwg Maybe it also fixes the issue described here: 
https://github.com/mixxxdj/mixxx/pull/4487
Can you have a look if this is likely?
